File: problems/tier1-foundation/006-discount-engine/boilerplate/go/part3/guided.go
package main

// Data class (given).
type CartItem struct {
	name string
	price float64
	quantity int
	category string
}

// HINT: introduce an abstraction so new rules don't change existing code.

// HINT: start from what this must return, then work backwards to the state it needs.
func apply_percentage_discount(cart []CartItem, percentage float64) float64 {
	// TODO: write your solution
	return 0.0
}

// HINT: start from what this must return, then work backwards to the state it needs.
func apply_flat_discount(cart []CartItem, amount float64) float64 {
	// TODO: write your solution
	return 0.0
}

// HINT: start from what this must return, then work backwards to the state it needs.
func apply_bogo(cart []CartItem, buyCount int, freeCount int) float64 {
	// TODO: write your solution
	return 0.0
}

// HINT: start from what this must return, then work backwards to the state it needs.
func apply_percentage_with_eligibility(cart []CartItem, percentage float64, minCartValue float64, requireFirstTimeUser bool, isFirstTimeUser bool, eligibleCategory string) float64 {
	// TODO: write your solution
	return 0.0
}

File: problems/tier1-foundation/006-discount-engine/boilerplate/javascript/part3/guided.js
// Data class (given).
class CartItem {
  constructor(name, price, quantity, category = "") {
    this.name = name;
    this.price = price;
    this.quantity = quantity;
    this.category = category;
  }
}

// HINT: introduce an abstraction so new rules don't change existing code.

// HINT: start from what this must return, then work backwards to the state it needs.
function apply_percentage_discount(cart, percentage) {
  // TODO: write your solution
  return null;
}

// HINT: start from what this must return, then work backwards to the state it needs.
function apply_flat_discount(cart, amount) {
  // TODO: write your solution
  return null;
}

// HINT: start from what this must return, then work backwards to the state it needs.
function apply_bogo(cart, buyCount, freeCount) {
  // TODO: write your solution
  return null;
}

// HINT: start from what this must return, then work backwards to the state it needs.
function apply_percentage_with_eligibility(cart, percentage, minCartValue, requireFirstTimeUser, isFirstTimeUser, eligibleCategory) {
  // TODO: write your solution
  return null;
}

// ── Export everything the test runner needs (do not remove) ──
// If you add classes (e.g. strategy subclasses), add them here too.
module.exports = { CartItem, apply_percentage_discount, apply_flat_discount, apply_bogo, apply_percentage_with_eligibility };
